### PR TITLE
Deprecate UnityPerCamera CBuffer

### PR DESCRIPTION
**Pull request by EvgeniiG**
Purpose of this PR:

Deprecate the UnityPerCamera CBuffer, which is set from C++, and replace it with UnityPerView, which is set entirely in C#.

It resolves issues with non-determinism (read-write conflicts?) affecting CBuffer variables when you render to several viewports at once (e.g. if you have several open tabs).

Additional objective is to improve the support of orthographic rendering, primarily concerning depth linearization.

This is a breaking change for the Shader Graph, which relies on UnityPerCamera. Previously, both LT and HD used UnityPerCamera, but now they behave differently. We need to decide how to handle this discrepancy in order to simplify the Shader Graph implementation.

Affected variable:
_Time (Rename: _CurrentTime, _PreviousTime, _DeltaTime, _SinCurrentTime, _CosCurrentTime), _ZBufferParams (Rename: _DepthBufferParams), _WorldSpaceCameraPos (rename _CameraPositionWS), _ScreenParams(rename _ScreenSize) and we don't use unity_OrthoParams

### CODE REVIEW

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl [hunk 1]
```diff
@@ -522,46 +522,55 @@ float3 LatlongToDirectionCoordinate(float2 coord)
...
 // Typically, this is the cheapest variant, provided you've already computed 'positionWS'.
-float LinearEyeDepth(float3 positionWS, float4x4 viewProjMatrix)
+// Assumes that the 'positionWS' is in front of the camera.
+float LinearEyeDepth(float3 positionWS, float4x4 viewMatrix)
 {
-    return mul(viewProjMatrix, float4(positionWS, 1.0)).w;
+    float viewSpaceZ = mul(viewMatrix, float4(positionWS, 1.0)).z;
+
+    // If the matrix is right-handed, we have to flip the Z axis to get a positive value.
+    return abs(viewSpaceZ);
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl]: Marker for disucssion :)

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl [hunk 2]
```diff
@@ -755,32 +765,35 @@ PositionInputs GetPositionInput(float2 positionSS, float2 invScreenSize, float d
...
     posInput.positionWS += depthOffsetVS * (-V);
+    posInput.deviceDepth = ComputeNormalizedDeviceCoordinatesWithZ(posInput.positionWS, viewProjMatrix).z;
 
-    float4 positionCS    = mul(viewProjMatrix, float4(posInput.positionWS, 1.0));
-    posInput.linearDepth = positionCS.w;
-    posInput.deviceDepth = positionCS.z / positionCS.w;
+    // Transform the displacement along the view vector to the displacement along the forward vector.
+    // Use abs() to make sure we get the sign right.
+    // 'depthOffsetVS' applies in the direction away from the camera.
+    posInput.linearDepth += depthOffsetVS * abs(dot(V, viewForwardDir));
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl]: Marker for discusison

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs [hunk 3]
```diff
@@ -15,13 +15,16 @@ public class HDCamera
         public Matrix4x4 viewMatrix;
         public Matrix4x4 projMatrix;
         public Matrix4x4 nonJitteredProjMatrix;
-        public Vector4 screenSize;
-        public Frustum frustum;
+        public Vector4   cameraPositionWS;
+        public float     detViewMatrix;
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs]: 16byte Alignment?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs [hunk 4]
```diff
@@ -396,23 +420,25 @@ public static void CleanUnused()
...
             cmd.SetGlobalMatrix(HDShaderIDs._NonJitteredViewProjMatrix, nonJitteredViewProjMatrix);
-            cmd.SetGlobalMatrix(HDShaderIDs._ViewProjMatrix, viewProjMatrix);
-            cmd.SetGlobalMatrix(HDShaderIDs._InvViewProjMatrix, viewProjMatrix.inverse);
-            cmd.SetGlobalVector(HDShaderIDs._ViewParam, viewParam);
-            cmd.SetGlobalVector(HDShaderIDs._InvProjParam, invProjParam);
-            cmd.SetGlobalVector(HDShaderIDs._ScreenSize, screenSize);
-            cmd.SetGlobalVector(HDShaderIDs._ScreenToTargetScale, scaleBias);
-            cmd.SetGlobalMatrix(HDShaderIDs._PrevViewProjMatrix, prevViewProjMatrix);
-            cmd.SetGlobalVectorArray(HDShaderIDs._FrustumPlanes, frustumPlaneEquations);
-            cmd.SetGlobalInt(HDShaderIDs._TaaFrameIndex, (int)taaFrameIndex);
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs]: Where is this? (_TaaFrameIndex)

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs [hunk 5]
```diff
@@ -533,9 +534,17 @@ public void PushGlobalParams(HDCamera hdCamera, CommandBuffer cmd, DiffusionProf
...
-                hdCamera.SetupGlobalParams(cmd);
-                if (m_FrameSettings.enableStereo)
-                    hdCamera.SetupGlobalStereoParams(cmd);
+                float  ct = Time.timeSinceLevelLoad;
+                float  pt = (m_TimeSinceLevelLoad > 0) ? m_TimeSinceLevelLoad : ct;
+                float  dt = Time.deltaTime;
+                float sdt = Time.smoothDeltaTime;
+
+                // Set up UnityPerFrame CBuffer.
+                cmd.SetGlobalVector(HDShaderIDs._CurrentTime,    new Vector4(ct * 0.05f, ct, ct * 2.0f, ct * 3.0f));
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: Add a a note that it is copy/paste from which file and based on unity builtin values

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs [hunk 6]
```diff
@@ -546,6 +555,13 @@ public void PushGlobalParams(HDCamera hdCamera, CommandBuffer cmd, DiffusionProf
                 var ssrefraction = VolumeManager.instance.stack.GetComponent<ScreenSpaceRefraction>()
                     ?? ScreenSpaceRefraction.@default;
                 ssrefraction.PushShaderParameters(cmd);
+
+                // Set up UnityPerView CBuffer.
+                hdCamera.SetupGlobalParams(cmd);
+                if (m_FrameSettings.enableStereo) hdCamera.SetupGlobalStereoParams(cmd);
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: We should setup stereo param in SetupGlobalParams instead of having it separate (can be in another PR)

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/ShaderVariables.hlsl [hunk 7]
```diff
@@ -24,13 +24,20 @@
...
 // ----------------------------------------------------------------------------
 
+//  *****************************************************
+//  *                                                   *
+//  *  UnityPerCamera has been deprecated. Do NOT use!  *
+//  *  Refer to UnityPerFrame and UnityPerView instead. *
+//  *                                                   *
+//  *****************************************************
+
 CBUFFER_START(UnityPerCamera)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/ShaderVariables.hlsl]: discussion marker - should we remove it ?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs [hunk 8]
```diff
@@ -211,15 +216,34 @@ public void Update(PostProcessLayer postProcessLayer, FrameSettings frameSetting
...
             {
                 Matrix4x4 cameraDisplacement = Matrix4x4.Translate(cameraPos - prevCameraPos); // Non-camera-relative positions
                 prevViewProjMatrix *= cameraDisplacement; // Now prevViewProjMatrix correctly transforms this frame's camera-relative positionWS
             }
 
-            frustum = Frustum.Create(viewProjMatrix, true, true);
+            float n = camera.nearClipPlane;
+            float f = camera.farClipPlane;
+
+            // Analyze the projection matrix.
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs]: Should we put this stuff in a helper function in CoreUtils? In the end we'll need that for LW as well if we want to unify everything for shader graph.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs [hunk 9]
```diff
@@ -396,23 +420,25 @@ public static void CleanUnused()
...
         }
 
+        // Set up UnityPerView CBuffer.
         public void SetupGlobalParams(CommandBuffer cmd)
         {
-            cmd.SetGlobalMatrix(HDShaderIDs._ViewMatrix, viewMatrix);
-            cmd.SetGlobalMatrix(HDShaderIDs._InvViewMatrix, viewMatrix.inverse);
-            cmd.SetGlobalMatrix(HDShaderIDs._ProjMatrix, projMatrix);
-            cmd.SetGlobalMatrix(HDShaderIDs._InvProjMatrix, projMatrix.inverse);
+            cmd.SetGlobalMatrix(HDShaderIDs._ViewMatrix,                viewMatrix);
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs]: As discussed, we might want to bring back OrthoParams/ProjectionParams even though they are not used right now (and maybe Forest demo was using them iirc?)

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs (hunk 6 above)
Comment (EvgeniiG) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: This code did not change, only formatting did.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs [hunk 10]
```diff
@@ -533,9 +534,17 @@ public void PushGlobalParams(HDCamera hdCamera, CommandBuffer cmd, DiffusionProf
...
-                if (m_FrameSettings.enableStereo)
-                    hdCamera.SetupGlobalStereoParams(cmd);
+                float  ct = Time.timeSinceLevelLoad;
+                float  pt = (m_TimeSinceLevelLoad > 0) ? m_TimeSinceLevelLoad : ct;
+                float  dt = Time.deltaTime;
+                float sdt = Time.smoothDeltaTime;
+
+                // Set up UnityPerFrame CBuffer.
+                cmd.SetGlobalVector(HDShaderIDs._CurrentTime,    new Vector4(ct * 0.05f, ct, ct * 2.0f, ct * 3.0f));
+                cmd.SetGlobalVector(HDShaderIDs._PreviousTime,   new Vector4(pt * 0.05f, pt, pt * 2.0f, pt * 3.0f));
```
Comment (JulienIgnace-Unity) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/HDRenderPipeline.cs]: Same as above, these might be useful as a helpers in CoreUtils for LW as well.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/ShaderVariablesFunctions.hlsl [hunk 11]
```diff
@@ -149,12 +149,7 @@ float3 GetViewForwardDir()
 // Returns 'true' if the current view performs a perspective projection.
 bool IsPerspectiveProjection()
 {
-#if defined(SHADERPASS) && (SHADERPASS != SHADERPASS_SHADOWS)
-    return (unity_OrthoParams.w == 0);
```
[the remark this answers is not in the document]
Comment (EvgeniiG) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/ShaderVariablesFunctions.hlsl]: No, we were using OrthoParams for efficiency reasons.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs (hunk 8 above)
Comment (EvgeniiG) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Camera/HDCamera.cs]: If Felipe wants us to. :-)

**Comment by sebastienlagarde** [comment]
Adding @phi-lira and @Kink3d for awareness.

Note: We change name of mentioned values:
_Time (Rename: _CurrentTime, _PreviousTime, _DeltaTime, _SinCurrentTime, _CosCurrentTime), _ZBufferParams (Rename: _DepthBufferParams), _WorldSpaceCameraPos (rename _CameraPositionWS), _ScreenParams(rename _ScreenSize) and we don't use unity_OrthoParams

to not conflict with C++ setup set in SetupCameraProperties. Long teram goal is to remove SetupCameraProperties, but for now in HD we need it for the postprocess as they are not part of SRP. This will change with V3 of PP that will be part of SRP project.
This also mean that currently nothing is broken with shader graph for HD (But shader graph will suffer from same rare bug of synchronisation than before this PR)

Pay special attention that we manage the time correctly with the "Animated material" flags of Editor (and better than builtin unity... i.e when disable time is 0 instead of varying for unknow reasons)
